v2/pkg/engine/plan/interface_selection_rewriter.go

```diff
@@ -0,0 +1,359 @@
+package plan
+
+import "github.com/wundergraph/graphql-go-tools/v2/pkg/ast"
+
+type interfaceSelectionRewriter struct {
+	operation  *ast.Document
+	definition *ast.Document
+
+	fieldRef        int
+	enclosingNode   ast.Node
+	dsConfiguration DataSourceConfiguration
+}
+
+func newInterfaceSelectionRewriter(operation *ast.Document, definition *ast.Document) *interfaceSelectionRewriter {
+	return &interfaceSelectionRewriter{
+		operation:  operation,
+		definition: definition,
+	}
+}
+
+func (r *interfaceSelectionRewriter) isFieldReturnsInterface(fieldRef int, enclosingNode ast.Node) (interfaceDefRef int, isInterface bool) {
+	var (
+		fieldDefRef int
+		hasField    bool
+	)
+
+	interfaceDefRef = ast.InvalidRef
+
+	switch enclosingNode.Kind {
+	case ast.NodeKindObjectTypeDefinition:
+		fieldDefRef, hasField = r.definition.ObjectTypeDefinitionFieldWithName(enclosingNode.Ref, r.operation.FieldNameBytes(fieldRef))
+		if !hasField {
+			return
+		}
+	case ast.NodeKindInterfaceTypeDefinition:
+		fieldDefRef, hasField = r.definition.InterfaceTypeDefinitionFieldWithName(enclosingNode.Ref, r.operation.FieldNameBytes(fieldRef))
+		if !hasField {
+			return
+		}
+	default:
+		return
+	}
+
+	fieldDefTypeName := r.definition.FieldDefinitionTypeNameBytes(fieldDefRef)
+	node, hasNode := r.definition.NodeByName(fieldDefTypeName)
+	if !hasNode {
+		return
+	}
+
+	if node.Kind != ast.NodeKindInterfaceTypeDefinition {
+		return
+	}
+
+	return node.Ref, true
+}
+
+func (r *interfaceSelectionRewriter) objectTypesImplementingInterface(interfaceDefRef int) (typeNames []string, ok bool) {
```

Comment:
this should be on ast.Document, I guess we already have it?

Reply:
no we don't, we have only `InterfaceTypeDefinitionImplementedByRootNodes` used on line 58 or we could check that particular object ref implements a particular interface

Reply:
will move to ast